packages/changed-elements-react/src/widgets/comparisonJobWidget/hooks/useNamedVersionLoader.tsx

```diff
@@ -0,0 +1,344 @@
+  if (currentFromNamedVersion)
+    return currentFromNamedVersion;
+  const currentFromChangeSet = getCurrentFromChangeSet(changeSets, currentChangeSetId);
+  if (currentFromChangeSet)
+    return currentFromChangeSet;
+  return {
+    id: currentChangeSetId,
+    displayName: IModelApp.localization.getLocalizedString("VersionCompare:versionCompare.currentChangeset"),
+    changesetId: currentChangeSetId,
+    changesetIndex: 0,
+    description: "",
+    createdDateTime: "",
+  };
+};
+
+const getCurrentFromNamedVersions = (namedVersions: NamedVersion[], currentChangeSetId: string, currentChangeSetIndex?: number) => {
+  const currentNamedVersion = namedVersions.find(version => (version.changesetId === currentChangeSetId || version.changesetIndex === currentChangeSetIndex));
+  if (currentNamedVersion) {
+    return currentNamedVersion;
+  }
+  return;
+};
+
+const getCurrentFromChangeSet = (changeSets: Changeset[], currentChangeSetId: string, currentChangeSetIndex?: number): NamedVersion | undefined => {
+  const currentChangeSet = changeSets.find(changeSet => (changeSet.id === currentChangeSetId || changeSet.index === currentChangeSetIndex));
+  if (currentChangeSet) {
+    return {
+      id: currentChangeSet.id,
+      displayName: currentChangeSet.displayName,
+      changesetId: currentChangeSet.id,
+      changesetIndex: currentChangeSet.index,
+      description: currentChangeSet.description,
+      createdDateTime: currentChangeSet.pushDateTime,
+    };
+  }
+  return;
+};
+
+const sortNamedVersions = (namedVersions: NamedVersion[], currentNamedVersion: NamedVersion, onError: () => void) => {
+  //if current index is 0 then no need to filter. All change sets are older than current.
+  const namedVersionsOlderThanCurrentVersion = currentNamedVersion.changesetIndex !== 0 ? namedVersions.filter(version => version.changesetIndex <= currentNamedVersion.changesetIndex) :
+    namedVersions;
+  if (namedVersionsOlderThanCurrentVersion.length === 0) {
+    onError();
+    return;
+  }
+  return namedVersionsOlderThanCurrentVersion.reverse();
+};
+
+type ProcessChangesetsArgs = {
+  iTwinId: string;
+  iModelId: string;
+  namedVersionLoaderState: NamedVersionLoaderState;
+  comparisonJobClient: IComparisonJobClient;
+  setResult: (result: NamedVersionLoaderResult) => void;
+  isDisposed: () => boolean;
+};
+
+type Entry = {
+  version: NamedVersion;
+  state: VersionProcessedState;
+  jobStatus: JobStatus;
+  jobProgress: JobProgress;
+};
+
+const processChangesetsAndUpdateResultState = async (args: ProcessChangesetsArgs) => {
+  const currentVersionId = args.namedVersionLoaderState.result.namedVersions.currentVersion.version.id;
+  // deletes all comparison jobs only for debugging purposes
+  //await Promise.all(args.namedVersionLoaderState.result.namedVersions.entries.map(async (entry) => {
+  //await deleteJob(args.comparisonJobClient, entry, args.iTwinId, args.iModelId, currentVersionId);
```

Comment:
I should remove them, but I kept them for now in case I have to demo again. It has been making it easier to show things off. I'll remove it on my last push.